### PR TITLE
Speedup environment activation

### LINKED ISSUE #13554: Spack Environments are painfully slow
I've been using Spack Environments for a while now, and so far my biggest complaint is the performance. On an NFS cluster in particular, it can take upwards of 15 minutes just to concretize an environment. Things are particularly bad when the environment is partially installed, as most of the time is spent updating the environment, not actually concretizing anything. I've even noticed that commands like `spack edit` or `spack help` will take orders of magnitude longer if I'm running them from a directory containing a `spack.yaml`. If I run `spack -d concretize`, I notice that we are acquiring and releasing locks on Spack's database thousands of times. Is there any way we can hold on to the lock until another process requests it so we don't need all this overhead?

@tgamblin

### PR DESCRIPTION
**Pull request by alalazo**
fixes #13554

### CODE REVIEW

**Comment by alalazo** [comment]
@adamjstewart Can you try this and report back? I'll try to see if I can improve it further in terms of performance.

**Comment by alalazo** [comment]
I think a refactor of `_set_variables_for_single_module` might also be a good thing, but given the entity of the performance bug reported in #13554 I'd be in  favor of merging this as a hotfix when considered done and do the refactor without hurry in a later PR.

**Comment by adamjstewart** [comment]
Here are the same benchmarks on my laptop. From my home directory:
```console
$ time spack help
...
real	0m2.366s
user	0m0.728s
sys	0m0.298s
```
From the directory containing the large `spack.yaml` (installed environment):
```console
$ time spack help
...
real	0m6.649s
user	0m4.290s
sys	0m1.054s
```
Definitely a drastic improvement over `develop`!

**Comment by alalazo** [comment]
@adamjstewart So far I tried to optimize for your small environment. Making a small break while the big one is building :slightly_smiling_face:

**Comment by alalazo** [comment]
@adamjstewart Tried on the big environment. I might shave something more if we are fine losing the information in:

https://github.com/spack/spack/blob/1114ae93752aa9d33c3520f72aa701d8105c09dd/lib/spack/spack/util/environment.py#L338-L348

in cases where we modify the environment a lot. If need be I can compute the information only in debug mode, but if the timings are good enough without this change I would keep the PR as it is. Let me know.

**Comment by adamjstewart** [comment]
I'm not sure how that function is used, so I don't know if we can skip it or not.

**Comment by alalazo** [comment]
@adamjstewart If there are warnings on e.g. the same variable set two times in a DAG (so the first set won't have any effect) it will print context to tell you where both set took place. It would be good to keep it as it is, so if the performance is acceptable right now I'd avoid making any change.

**Comment by adamjstewart** [comment]
Yeah, I guess it depends on how much of a difference it makes. The current PR drops us from 2 minutes to 6 seconds. If it would drop us from 6 seconds to 3 seconds, it would be worth it in my mind.

**Comment by alalazo** [comment]
> The current PR drops us from 2 minutes to 6 seconds.

You didn't try the latest 2 commits :wink:

**Comment by adamjstewart** [comment]
Okay, with the latest commit, I'm down to 4 seconds.

File: lib/spack/spack/build_environment.py [hunk 1]
```diff
@@ -489,19 +512,37 @@ def static_to_shared_library(static_lib, shared_lib=None, **kwargs):
...
     """
     # If a user makes their own package repo, e.g.
     # spack.pkg.mystuff.libelf.Libelf, and they inherit from an existing class
     # like spack.pkg.original.libelf.Libelf, then set the module variables
     # for both classes so the parent class can still use them if it gets
     # called. parent_class_modules includes pkg.module.
     modules = parent_class_modules(pkg.__class__)
+    if context not in ('run', 'build'):
+        msg = 'context must be either "run" or "build"'
+        raise RuntimeError(msg)
```
Comment (tgamblin) [inline lib/spack/spack/build_environment.py]: This should really be a `ValueError`.

**Comment by alalazo** [comment]
@tgamblin I reverted c2f6113 and implemented request number 1. Without that commit we retain most of the speedup so it makes sense also to me to look for further improvements later.

**Comment by tgamblin** [comment]
@adamjstewart: how fast is this for you now?

**Comment by adamjstewart** [comment]
On my laptop:

### Before (develop)

From my home directory:
```console
$ time spack help
...
real	0m2.871s
user	0m0.719s
sys	0m0.307s
```
From a directory containing a large `spack.yaml`:
```console
$ time spack help
...
real	3m22.219s
user	3m10.927s
sys	0m3.275s
```

### After (fixes/speedup_environment_activation)

From my home directory:
```console
$ time spack help
...
real	0m1.972s
user	0m0.682s
sys	0m0.271s
```
From a directory containing a large `spack.yaml`:
```console
$ time spack help
...
real	0m5.015s
user	0m3.607s
sys	0m0.612s
```

Let me know if you want benchmarks on Blue Waters where the problem is even more pronounced.

**Pull request merged**